query.go

```diff
@@ -334,14 +336,15 @@ func (q *query) terminate(ctx context.Context, reason LookupTerminationReason) {
 			Key:       q.key,
 			Terminate: &LookupTerminateEvent{Reason: reason},
 		})
+		cancel() // abort outstanding queries
 		q.terminated = true
 	}
 }
 
 // queryPeer queries a single peer and reports its findings on the channel.
 // queryPeer does not access the query state in queryPeers!
-func (q *query) queryPeer(ch chan<- *queryUpdate, p peer.ID) {
-	dialCtx, queryCtx := q.ctx, q.ctx
+func (q *query) queryPeer(ctx context.Context, ch chan<- *queryUpdate, p peer.ID) {
+	dialCtx, queryCtx := ctx, ctx
```

Comment:
Doing this will mean that we are likely to end up redialing peers we just cancelled since in the lookup followup we'll query the top K peers that we have yet to query.